### PR TITLE
IH-689: Autogenerate record_util.ml (where it's reasonable)

### PR DESCRIPTION
**Pull request by last-genius**
Best reviewed by commit - commit messages include the necessary context.

Changes can mostly be grouped into:
* Implementing autogeneration
* Removing manually-specified conversion functions that are identical to the automatically-generated ones
* Removing converters identical in function but not in form **(requires renaming usage sites)** - these are commit-per-function for ease of review, but can be squashed before merging
* Removing conversion functions that were not covered with tests **(requires also adding corresponding tests)** - these are commit-per-function
* Justifying the grim shadowing existence of the rest.

It might seem like a big patch, but most of it is removing and renaming things, with only a tiny bit of new code. Since Edwin's PR some time ago (#5349), we can now confirm the behaviour of the autogenerated code is the same as the old code, too!

It's better to autogenerate `record_util.ml` because we already have a lot of boilerplate that needs to be touched to add new things to the datamodel and CLI, this reduces the boring parts. In addition, the manual nature of this file caused duplicates (that existed since before git! And nobody noticed!), inconsistent behaviour (dots/dashes/underscores/case-sensitivity), odd cases (functions converting variants **not** in the datamodel), etc. The sooner we can get rid of the large part of this the better.

### CODE REVIEW

File: ocaml/idl/ocaml_backend/gen_api.ml [hunk 1]
```diff
@@ -354,6 +354,38 @@ let toposort_types highapi types =
   assert (List.sort compare result = List.sort compare types) ;
   result
 
+let gen_utils highapi =
```
Comment (psafont) [inline ocaml/idl/ocaml_backend/gen_api.ml:357]: Utils is not very useful name, maybe `gen_record_deser`?

**Comment by psafont** [comment]
I really like the approach used here to keep maintenance of polymorphic variants in the datamodel to a minimum.

About the VM powerstate I'll wait for @robhoes to reply as he will know more about the history that needed the two additional states

File: ocaml/xapi/xapi_vm_clone.ml [hunk 2]
```diff
@@ -231,6 +231,8 @@ let quiesced = "quiesced"
 
 let snapshot_info ~power_state ~is_a_snapshot =
   let power_state_info =
+    (* Uses a shadowed function from record_util because xapi_vm_clone.mli
+       interface includes non-datamodel enum variants *)
```
Comment (minglumlu) [inline ocaml/xapi/xapi_vm_clone.ml:235]: I can't find where the non-datamodel enum variants are used. Are they really useful?

**Comment by psafont** [comment]
Just in time for their 18th birthday...

**Comment by last-genius** [comment]
Addressed the review comments. XenRT tests failed all over the place though, will re-run and try to find some time this week to look into them. Converting to draft for now.

**Comment by robhoes** [comment]
A good simplification of the code, removing lots of lines. I can't see anything wrong with this.

**Comment by last-genius** [comment]
@robhoes pointed out that XenRT, beside our own tests, rely on the string output of vm_power_state conversions to be lowercase - this is likely the cause for suite failures. I've removed two deprecated power states, and kept the lowercase function (and its users) as is.

File: ocaml/xapi-cli-server/record_util.ml [hunk 3]
```diff
@@ -801,6 +786,8 @@ let string_to_vbd_type s =
   | _ ->
       record_failure "Expected 'CD' or 'Disk', got %s" s
 
+(* Intentional shadowing - some usage sites rely on the
+   non-capitalized string output of the conversion function. *)
 let power_to_string h =
```
[the remark this answers is not in the document]
Comment (last-genius) [reply ocaml/xapi-cli-server/record_util.ml]: Made this change.

**Review by robhoes — APPROVED** [review]

**Review by psafont — APPROVED** [review]

**Pull request merged**